File: 프로그래머스/lv2/76502. 괄호 회전하기/괄호 회전하기.js
const openBracket = ['(', '{', '['];

function solution(s) {
    let count = 0;
    
    for(let i = 0; i < s.length; i++) {
        // 문자열 자르는 코드 필요
        const str = s.substr(1, s.length-1);
        s = str+s[0];
        if(isValid(s)) count++;
    }
    return count;
}

// 괄호쌍이 맞는다면 
const isValid = str => {
    const stack = [];
    
    for(let i = 0; i < str.length; i++) {
        // 열린 괄호
        
        if(openBracket.includes(str[i])) stack.push(str[i]);
        else {
            if(stack.length === 0) return false;
            else if(stack[stack.length - 1] === '{' && str[i] === '}') stack.pop();  
            else if(stack[stack.length - 1] === '[' && str[i] === ']') stack.pop();  
            else if(stack[stack.length - 1] === '(' && str[i] === ')') stack.pop();  
        }
    }
    return stack.length === 0;
}
/*
주어진 문자열을 문자열 길이만큼 회전시키는 반복을 한다. 
올바른 괄호임을 판별하는 함수가 필요하다.
*/

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
from collections import deque

def rotate(s):
    new_s = deque(list(s))
    
    left = new_s.popleft()
    new_s.append(left)
    
    return ''.join(new_s)
    
def is_valid(s):
    r = deque(list(s))
    l = []
    
    for i in range(len(s)):
        char = r[0]
        
        if i == 0:
            if char == '[' or char == '{' or char == '(':
                r.popleft()
                l.append(char)
            else:
                return False
            
        else:    
            if char == '[' or char == '{' or char == '(':
                r.popleft()
                l.append(char)
            
            else:
                if len(l) <= 0:
                    return False
                else:
                    lt = l[-1]
                    if lt == '[' and char == ']':
                        r.popleft()
                        l.pop()
                    elif lt == '{' and char == '}':
                        r.popleft()
                        l.pop()
                    elif lt == '(' and char == ')':
                        r.popleft()
                        l.pop()
                    else:
                        return False
                    
    if len(r) != 0 or len(l) != 0:
        return False

    return True
    
def solution(s):
    temp_s = s
    count = 0
    
    for i in range(len(s) - 1):
        if i == 0:
            if is_valid(temp_s):
                count += 1
        
        else:
            temp_s = rotate(temp_s)
        
            if is_valid(temp_s):
                count += 1
            
    return count
        
        
